Tensile/KernelWriterAssembly.py

```diff
@@ -5211,30 +5208,237 @@ def lraFinalOffset(self, kernel, tP):
+            # calculate WaveFrontLAneoffset
+            kStr += inst("v_and_b32", \
+                vgpr(rReg), \
+                hex(globalParameters["WavefrontWidth"]-1), \
+                vgpr(dividendReg), \
+                "WaveLaneoffset =  %s & %s-1"%(vgpr(dividendReg),hex(globalParameters["WavefrontWidth"])))
+            kStr += inst("v_lshrrev_b32", \
+                vgpr(rReg), \
+                log2(kernel["MatrixInstM"]), \
+                vgpr(rReg), \
+                "M_KIdx = WaveLaneoffset << %s"%(log2(kernel["MatrixInstM"])))
+            if kernel["ProblemType"]["DataType"].isBFloat16() :
+              #adjust Kindex value for bfloat16
+              kStr += inst("v_lshlrev_b32", \
+                  vgpr(rReg), \
+                  log2(self.bpr//tP["bpe"]), \
+                  vgpr(rReg), \
+                  "M_KIdx = v1 %% %s*%s"%(log2(kernel["MatrixInstM"]),tP["bpe"]))
+            if kernel["ProblemType"]["DataType"].isHalf():
+              #adjust Kindex value for half
+              # numSrcs =2
+              kStr += inst("v_lshlrev_b32", \
+                  vgpr(rReg), \
+                  log2(2*self.bpr//tP["bpe"]), \
+                  vgpr(rReg), \
+                  "M_KIdx = v1 %% %s*%s"%(log2(kernel["MatrixInstM"]),tP["bpe"]))
+            kStr += inst("s_mov_b32", \
+                sgpr(tmpSgpr), \
+                hex(kernel["MacroTile0"]), \
+                "")
+            kStr += inst("v_mul_lo_u32", \
+                vgpr(rReg), \
+                sgpr(tmpSgpr), \
+                vgpr(rReg), \
+                "M_Koffset = M_Kidx+MT0")
+            kStr += inst("v_and_b32", \
+                vgpr(tP["gpr"]["lro"]), \
+                hex(kernel["MatrixInstM"]-1), \
+                vgpr(tP["gpr"]["lro"]), \
+                "Mi_Kidx = LRO & %s-1"%(hex(kernel["MatrixInstM"])))
+            kStr += inst("v_add_u32", \
+                vgpr(tP["gpr"]["lro"]), \
+                vgpr(rReg),\
+                vgpr(tP["gpr"]["lro"]), \
+                "M_Koffset = Mi_Kidx + M_Koffset")
+            kStr += inst("v_add_u32", \
+                vgpr(tP["gpr"]["lro"]), \
+                vgpr(tmpVgpraddr),\
+                vgpr(tP["gpr"]["lro"]), \
+                "LRO = LRO + M_Koffset")
+          else:
+            kStr += inst("v_and_b32", \
+                vgpr(rReg), \
+                hex(globalParameters["WavefrontWidth"]-1), \
+                vgpr(dividendReg), \
+                "WaveLaneoffset =  %s & %s-1"%(vgpr(dividendReg),hex(globalParameters["WavefrontWidth"])))
+            #optimization : fetch both MiK+0, MiK+B in single local Read
+            kStr += inst("v_lshrrev_b32", \
+                vgpr(rReg), \
+                log2(kernel["MatrixInstM"]), \
+                vgpr(rReg), \
+                "N_KIdx = v1 >> %s"%(log2(kernel["MatrixInstM"])))
+            if kernel["ProblemType"]["DataType"].isBFloat16() :
```

Comment:
16-bit code is duplicated in a few places, could be consolidated a bit

Reply:
done.